File: src/Software Engineer Prepare Kit/10/solution.ts
class SinglyLinkedListNode {
	data: number
	next: SinglyLinkedListNode | null

	constructor(nodeData: number) {
		this.data = nodeData
		this.next = null
	}
}

class SinglyLinkedList {
	head: SinglyLinkedListNode | null
	tail: SinglyLinkedListNode | null

	constructor() {
		this.head = null
		this.tail = null
	}

	insertNode(nodeData: number): void {
		const node = new SinglyLinkedListNode(nodeData)

		if (this.head == null) {
			this.head = node
		} else {
			this.tail!.next = node
		}

		this.tail = node
	}
}

export default function removeKthNodeFromEnd(head: SinglyLinkedListNode, k: number): SinglyLinkedListNode {
	const dummy = new SinglyLinkedListNode(0)
	dummy.next = head

	let slow = dummy
	let fast: SinglyLinkedListNode | null = dummy

	let i = 0
	while (i < k + 2) {
		if (fast === null) return head

		fast = fast.next
		i++
	}

	while (fast !== null) {
		slow = slow.next!
		fast = fast.next
	}

	slow.next = slow.next!.next

	return dummy.next
}

export { SinglyLinkedListNode, SinglyLinkedList }

File: src/Software Engineer Prepare Kit/10/solution.go
package main

import (
	"fmt"
)

type SinglyLinkedListNode struct {
	data int32
	next *SinglyLinkedListNode
}

type SinglyLinkedList struct {
	head *SinglyLinkedListNode
	tail *SinglyLinkedListNode
}

func (singlyLinkedList *SinglyLinkedList) insertNodeIntoSinglyLinkedList(nodeData int32) {
	node := &SinglyLinkedListNode{
		next: nil,
		data: nodeData,
	}

	if singlyLinkedList.head == nil {
		singlyLinkedList.head = node
	} else {
		singlyLinkedList.tail.next = node
	}

	singlyLinkedList.tail = node
}

func printSinglyLinkedList(node *SinglyLinkedListNode, sep string) {
	for node != nil {
		fmt.Printf("%d", node.data)

		node = node.next

		if node != nil {
			fmt.Printf("%s", sep)
		}
	}
}

func RemoveKthNodeFromEnd(head *SinglyLinkedListNode, k int32) *SinglyLinkedListNode {
	dummy := &SinglyLinkedListNode{next: head}
	slow := dummy
	fast := dummy

	for range k + 2 {
		if fast == nil {
			return head
		}

		fast = fast.next
	}

	for fast != nil {
		slow = slow.next
		fast = fast.next
	}

	slow.next = slow.next.next

	return dummy.next
}
